Add other metadata files to 4.1.3 (#464)
OWASP/wstg: document/4-Web_Application_Security_Testing/01-Information_Gathering/03-Review_Webserver_Metafiles_for_Information_Leakage.md

2. Create the list of directories that are to be avoided by Spiders, Robots, or Crawlers.
3. Identify other information which may benefit testers or attackers.

## How to Test

> Any of the actions performed below with `wget` could also be done with `curl`. Many Dynamic Application Security Testing (DAST) tools such as ZAP and Burp Suite include checks or parsing for these resources as part of their spider/crawler functionality. They can also be identified using various [Google Dorks](https://en.wikipedia.org/wiki/Google_hacking) or leveraging advanced search features such as `inurl:`.

### Robots

Web Spiders, Robots, or Crawlers retrieve a web page and then recursively traverse hyperlinks to retrieve further web content. Their accepted behavior is specified by the [Robots Exclusion Protocol](https://www.robotstxt.org) of the [robots.txt](https://www.robotstxt.org/) file in the web root directory.

As an example, the beginning of the `robots.txt` file from [Google](https://www.google.com/robots.txt) sampled on 2020 May 5 is quoted below:

```text
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
Disallow: /sdch
...
```

> The [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) directive refers to the specific web spider/robot/crawler. For example, the `User-Agent: Googlebot` refers to the spider from Google while `User-Agent: bingbot` refers to crawler from Microsoft. `User-Agent: *` in the example above applies to all [web spiders/robots/crawlers](https://support.google.com/webmasters/answer/6062608?visit_id=637173940975499736-3548411022&rd=1).

Reviewer: `refers to crawler from Microsoft` This sounds wrong no? Maybe `to a`?

The `robots.txt` file is retrieved from the web root directory of the web server. For example, to retrieve the `robots.txt` from `www.google.com` using `wget` or `curl`:

```bash
$ curl -O -Ss http://www.google.com/robots.txt && head -n5 robots.txt
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
...
```

#### Analyze robots.txt Using Google Webmaster Tools

Web site owners can use the Google "Analyze robots.txt" function to analyze the website as part of its [Google Webmaster Tools](https://www.google.com/webmasters/tools). This tool can assist with testing and the procedure is as follows:

1. Sign into Google Webmaster Tools with a Google account.
2. On the dashboard, enter the URL for the site to be analyzed.
3. Choose between the available methods and follow the on screen instruction.

### META Tags

`<META>` tags are located within the `HEAD` section of each HTML document and should be consistent across a web site in the event that the robot/spider/crawler start point does not begin from a document link other than webroot i.e. a [deep link](https://en.wikipedia.org/wiki/Deep_linking).

#### Robots META Tag

> 

Reviewer: A link in here to this would be good: https://www.robotstxt.org/meta.html

The following example was retrieved from Google 2020 May 05:

```bash
$ wget --no-verbose  https://www.google.com/humans.txt && cat humans.txt
2020-05-07 12:57:52 URL:https://www.google.com/humans.txt [286/286] -> "humans.txt" [1]
Google is built by a large team of engineers, designers, researchers, robots, and others in many different sites across the globe. It is updated continuously, and built with more tools and technologies than we can shake a stick at. If you'd like to help us out, see careers.google.com.
```

### Other .well-known Information Sources

There are other RFCs and Internet drafts which suggest standardized uses of files within the `.well-known/` directory. Lists of which can be found [here](https://en.wikipedia.org/wiki/List_of_/.well-known/_services_offered_by_webservers) or [here](https://www.iana.org/assignments/well-known-uris/well-known-uris.xhtml).

> If would be fairly simple for a tester to review the RFC/drafts are create a list to be supplied to a crawler or fuzzer, in order to verify the existence or content of such files.

Reviewer: `It` instead of `If`?

2. Create the list of directories that are to be avoided by Spiders, Robots, or Crawlers.
3. Identify other information which may benefit testers or attackers.

## How to Test

> Any of the actions performed below with `wget` could also be done with `curl`. Many Dynamic Application Security Testing (DAST) tools such as ZAP and Burp Suite include checks or parsing for these resources as part of their spider/crawler functionality. They can also be identified using various [Google Dorks](https://en.wikipedia.org/wiki/Google_hacking) or leveraging advanced search features such as `inurl:`.

### Robots

Web Spiders, Robots, or Crawlers retrieve a web page and then recursively traverse hyperlinks to retrieve further web content. Their accepted behavior is specified by the [Robots Exclusion Protocol](https://www.robotstxt.org) of the [robots.txt](https://www.robotstxt.org/) file in the web root directory.

As an example, the beginning of the `robots.txt` file from [Google](https://www.google.com/robots.txt) sampled on 2020 May 5 is quoted below:

```text
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
Disallow: /sdch
...
```

> The [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) directive refers to the specific web spider/robot/crawler. For example, the `User-Agent: Googlebot` refers to the spider from Google while `User-Agent: bingbot` refers to crawler from Microsoft. `User-Agent: *` in the example above applies to all [web spiders/robots/crawlers](https://support.google.com/webmasters/answer/6062608?visit_id=637173940975499736-3548411022&rd=1) as quoted below:

Reviewer: ```suggestion
The [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) directive refers to the specific web spider/robot/crawler. For example, the `User-Agent: Googlebot` refers to the spider from Google while `User-Agent: bingbot` refers to crawler from Microsoft. `User-Agent: *` in the example above applies to all [web spiders/robots/crawlers](https://support.google.com/webmasters/answer/6062608?visit_id=637173940975499736-3548411022&rd=1).
```

> `User-agent: *`

Reviewer: this line can be removed.

2. On the dashboard, enter the URL for the site to be analyzed.
3. Choose between the available methods and follow the on screen instruction.

### META Tags

`<META>` tags are located within the `HEAD` section of each HTML document and should be consistent across a web site in the event that the robot/spider/crawler start point does not begin from a document link other than webroot i.e. a [deep link](https://en.wikipedia.org/wiki/Deep_linking).

#### Robots Meta Tag

If there is no `<META NAME="ROBOTS" ... >` entry then the "Robots Exclusion Protocol" defaults to `INDEX,FOLLOW` respectively. Therefore, the other two valid entries defined by the “Robots Exclusion Protocol” are prefixed with `NO...` i.e. `NOINDEX` and `NOFOLLOW`.

##### Robots META Tags - with Burp

Based on the Disallow directive(s) listed within the `robots.txt` file in webroot, a regular expression search for `<META NAME="ROBOTS"` within each web page is undertaken and the result compared to the `robots.txt` file in webroot.

For example, the `robots.txt` file from `facebook.com` has a `Disallow: /album.php` and the resulting search for `<META NAME="ROBOTS"` is shown below:

> ![Facebook Meta Tag Example](images/Meta_Tag_Example-Facebook-Aug_2013.png)\

Reviewer: Can we possibly replace this image? It doesn't show neither one of the 4 entries you discussed in the previous section, and it's just a bit old? Doesn't need to be part of a Burp request quite honestly.

Author: I'll review/revise. I did validate that the content is still the same as 2013...

```text
...
Disallow: /search
...
Disallow: /sdch
...
```

Web spiders/robots/crawlers can [intentionally ignore](https://blog.isc2.org/isc2_blog/2008/07/the-attack-of-t.html) the `Disallow` directives specified in a `robots.txt` file, such as those from [Social Networks](https://www.htbridge.com/news/social_networks_can_robots_violate_user_privacy.html) to ensure that shared linked are still valid. Hence, `robots.txt` should not be considered as a mechanism to enforce restrictions on how web content is accessed, stored, or republished by third parties.

The `robots.txt` file is retrieved from the web root directory of the web server. For example, to retrieve the `robots.txt` from `www.google.com` using `wget` or `curl`:

```bash
$ wget https://www.google.com/robots.txt
--2020-05-05 08:33:43--  https://www.google.com/robots.txt
Resolving www.google.com (www.google.com)... 172.217.13.164, 2607:f8b0:4020:805::2004
Connecting to www.google.com (www.google.com)|172.217.13.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/plain]
Saving to: ‘robots.txt’

> robots.txt                                 [ <=>                                                                        ]   6.84K  --.-KB/s    in 0.001s  

Author: Sure, it was a literal copy/paste from a terminal window. I didn’t put any consideration into reformatting it....

> robots.txt                                 [ <=>                                                                        ]   6.84K  --.-KB/s    in 0.001s  

Author: Sure, it was a literal copy/paste from a terminal window. I didn’t put any consideration into reformatting it....

## Summary

This section describes how to test various metadata files for information leakage of the web application's path(s), or functionality. Furthermore, the list of directories that are to be avoided by Spiders, Robots, or Crawlers can also be created as a dependency for [Map execution paths through application](07-Map_Execution_Paths_Through_Application.md). Other information may also be collected to identify attack surface, technology details, or for use in social engineering engagement.

## Test Objectives

1. Information leakage of the web application's path(s) or functionality.
2. Create the list of directories that are to be avoided by Spiders, Robots, or Crawlers.
3. Identify other information which may benefit testers or attackers.

## How to Test

Note: Any of the actions performed below with `wget` could also be done with `curl`. Many Dynamic Application Security Testing (DAST) tools such as ZAP and Burp Suite include checks or parsing for these resources as part of their spider/crawler functionality. They can also be identified using various [Google Dorks](https://en.wikipedia.org/wiki/Google_hacking) or leveraging advanced search features such as `inurl:`.

> 

Author: I'm not sure what you're asking about here.

Are you wanting to establish a default, or have I not used what you feel is the default?

> 

Reviewer: Notes are usually set in `> `.
So you'd drop the `Note:` and replace it with a `> `. It makes the text automatically stand out and actually look like a note.

### Robots

Web Spiders, Robots, or Crawlers retrieve a web page and then recursively traverse hyperlinks to retrieve further web content. Their accepted behavior is specified by the [Robots Exclusion Protocol](https://www.robotstxt.org) of the [robots.txt](https://www.robotstxt.org/) file in the web root directory.

> As an example, the beginning of the `robots.txt` file from <https://www.google.com/robots.txt> sampled on 2020 May 5 is quoted below:

Reviewer 2: I think we have removed all link representation like `<https://www.google.com/robots.txt>`. Can we change that ?

Author: @ThunderSon ?

I'm with @rejahrehim I don't like this format either, but I didn't really have a reason to disagree. Other than it 'looks' horribly HTML'ish and we were trying not to mix HTML into our MD.

Author: Shall  we just go like this?
```suggestion
As an example, the beginning of the `robots.txt` file from [Google](https://www.google.com/robots.txt) sampled on 2020 May 5 is quoted below:
```

Reviewer: Yeah, that looks better. I just don't really enjoy having links repeated like that. Not really the way to use that format. This is not HTML'ish, it md:
> URLs and URLs in angle brackets will automatically get turned into links.
It doesn't bother me either ways.

Author: That's why I said "ish" :wink:

```text
...
Disallow: /search
...
Disallow: /sdch
...
```

Web spiders/robots/crawlers can [intentionally ignore](https://blog.isc2.org/isc2_blog/2008/07/the-attack-of-t.html) the `Disallow` directives specified in a `robots.txt` file, such as those from [Social Networks](https://www.htbridge.com/news/social_networks_can_robots_violate_user_privacy.html) to ensure that shared linked are still valid. Hence, `robots.txt` should not be considered as a mechanism to enforce restrictions on how web content is accessed, stored, or republished by third parties.

The `robots.txt` file is retrieved from the web root directory of the web server. For example, to retrieve the `robots.txt` from `www.google.com` using `wget` or `curl`:

```bash
$ wget https://www.google.com/robots.txt
--2020-05-05 08:33:43--  https://www.google.com/robots.txt
Resolving www.google.com (www.google.com)... 172.217.13.164, 2607:f8b0:4020:805::2004
Connecting to www.google.com (www.google.com)|172.217.13.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/plain]
Saving to: ‘robots.txt’

robots.txt
[ <=>                         ]   6.84K  --.-KB/s    in 0.001s  

> 2020-05-05 08:33:43 (9.89 MB/s) - ‘robots.txt’ saved [7004]

Reviewer 3: All the extra wget and curl output seems like it's just wasted space - could it be removed?

Author: I'd be up for trimming the first example to just use curl (instead of both), but for new testers theres value in seeingbactual examples to know what to expect etc.

Reviewer 3: If you want real output, you could add `-Ss` to curl, which hides most of the output unless there's errors?

User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
$
```

#### Analyze robots.txt Using Google Webmaster Tools

Web site owners can use the Google "Analyze robots.txt" function to analyze the website as part of its [Google Webmaster Tools](https://www.google.com/webmasters/tools). This tool can assist with testing and the procedure is as follows:

1. Sign into Google Webmaster Tools with a Google account.
2. On the dashboard, enter the URL for the site to be analyzed.
3. Choose between the available methods and follow the on screen instruction.

### META Tag

`<META>` tags are located within the `HEAD` section of each HTML document and should be consistent across a web site in the event that the robot/spider/crawler start point does not begin from a document link other than webroot i.e. a [deep link](https://en.wikipedia.org/wiki/Deep_linking).

If there is no `<META NAME="ROBOTS" ... >` entry then the "Robots Exclusion Protocol" defaults to `INDEX,FOLLOW` respectively. Therefore, the other two valid entries defined by the “Robots Exclusion Protocol” are prefixed with `NO...` i.e. `NOINDEX` and `NOFOLLOW`.

> Web spiders/robots/crawlers can intentionally ignore the `<META NAME="ROBOTS"` tag as the `robots.txt` file convention is preferred. Hence, **<META> Tags should not be considered the primary mechanism, rather a complementary control to robots.txt**.

Reviewer 3: I disagree with this. Bots can ignore `robots.txt` just as easily as they can ignore `<meta>` tags, and the latter approach has the advantage of not revealing al the pages that you don't want indexed. Tags should be used for any pages that you don't want to advertise the existence of.

Google also [say](https://support.google.com/webmasters/answer/6062608?hl=en) that they will ignore `robots.txt` if there's a direct link to the page, but will honour the `<meta>` tag:

> **A robotted page can still be indexed if linked to from from other sites**
>  While Google won't crawl or index the content blocked by robots.txt, we might still find and index a disallowed URL if it is linked from other places on the web. As a result, the URL address and, potentially, other publicly available information such as anchor text in links to the page can still appear in Google search results. **To properly prevent your URL from appearing in Google Search results, you should password-protect the files on your server or use the noindex meta tag** or response header (or remove the page entirely).

Author: I agree with your points. I'm just going to remove the bold/note text as this isn't a dev/implementation guide.

Reviewer 3: I'd be tempted to also remove:

> Web spiders/robots/crawlers can intentionally ignore the <META NAME="ROBOTS" tag as the robots.txt file convention is preferred

It implies that they can't intentionally ignore robots.txt (which they can) - and also would be a developer recommendation (which I'd disagree with as per the above)

### Sitemaps

A sitemap is a file where a developer or organization can provide information about the pages, videos, and other files offered by the site or application, and the relationship between them. Search engines can use this file to more intelligently explore your site. Testers can use `sitemap.xml` files to learn more about the site or application to explore it more completely.

The following excerpt is from Google's primary sitemap retrieved 2020 May 05.

```bash
$ wget https://www.google.com/sitemap.xml && head -n8 sitemap.xml
--2020-05-05 12:23:30--  https://www.google.com/sitemap.xml
Resolving www.google.com (www.google.com)... 172.217.13.164, 2607:f8b0:4020:806::2004
Connecting to www.google.com (www.google.com)|172.217.13.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2049 (2.0K) [text/xml]
Saving to: ‘sitemap.xml.3’

sitemap.xml.3
100%[=============================================>]   2.00K  --.-KB/s    in 0s

2020-05-05 12:23:30 (9.24 MB/s) - ‘sitemap.xml.3’ saved [2049/2049]

> <?xml version="1.0" encoding="UTF-8"?>

Reviewer 3: Might look nicer with XML syntax highlighting?

Author: Will test options.

Author: I left the initial one as bash since it really is totally terminal output. The second one I trimmed off the wget output and left just the xml and switch to xml hint.

Reviewer 3: Yeah, that looks better.

## How to Test

> Note: Any of the actions performed below with `wget` could also be done with `curl`. Many Dynamic Application Security Testing (DAST) tools such as ZAP and Burp Suite include checks or parsing for these resources as part of their spider/crawler functionality. They can also be identified using various 'Google Dorks' or leveraging advanced search features such as `inurl:`.

Reviewer: ```suggestion
Note: Any of the actions performed below with `wget` could also be done with `curl`. Many Dynamic Application Security Testing (DAST) tools such as ZAP and Burp Suite include checks or parsing for these resources as part of their spider/crawler functionality. They can also be identified using various [Google Dorks](https://en.wikipedia.org/wiki/Google_hacking) or leveraging advanced search features such as `inurl:`.
```

### robots.txt

Web Spiders, Robots, or Crawlers retrieve a web page and then recursively traverse hyperlinks to retrieve further web content. Their accepted behavior is specified by the [Robots Exclusion Protocol](https://www.robotstxt.org) of the [robots.txt](https://www.robotstxt.org/) file in the web root directory.

> As an example, the beginning of the robots.txt file from [https://www.google.com/robots.txt](https://www.google.com/robots.txt) sampled on 2020 May 5 is quoted below:

Reviewer: ```suggestion
As an example, the beginning of the `robots.txt` file from <https://www.google.com/robots.txt> sampled on 2020 May 5 is quoted below:
```

```text
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks
Disallow: /sdch
...
```

> The *User-Agent* directive refers to the specific web spider/robot/crawler. For example the *User-Agent: Googlebot* refers to the spider from Google while “User-Agent: bingbot” refers to crawler from Microsoft/Yahoo!. `User-Agent: *` in the example above applies to all [web spiders/robots/crawlers](https://support.google.com/webmasters/answer/6062608?visit_id=637173940975499736-3548411022&rd=1) as quoted below:

Reviewer: ```suggestion
The [User-Agent](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) directive refers to the specific web spider/robot/crawler. For example, the `User-Agent: Googlebot` refers to the spider from Google while `User-Agent: bingbot` refers to crawler from Microsoft. `User-Agent: *` in the example above applies to all [web spiders/robots/crawlers](https://support.google.com/webmasters/answer/6062608?visit_id=637173940975499736-3548411022&rd=1) as quoted below:
```

`User-agent: *`

The `Disallow` directive specifies which resources are prohibited by spiders/robots/crawlers. In the example above, the following are prohibited:

```text
...
Disallow: /search
...
Disallow: /sdch
...
```

Web spiders/robots/crawlers can [intentionally ignore](https://blog.isc2.org/isc2_blog/2008/07/the-attack-of-t.html) the `Disallow` directives specified in a robots.txt file, such as those from [Social Networks](https://www.htbridge.com/news/social_networks_can_robots_violate_user_privacy.html) to ensure that shared linked are still valid. Hence, robots.txt should not be considered as a mechanism to enforce restrictions on how web content is accessed, stored, or republished by third parties.

> #### robots.txt in Webroot - with wget or curl

Reviewer: What does this title even mean?

The robots.txt file is retrieved from the web root directory of the web server. For example, to retrieve the robots.txt from `www.google.com` using `wget` or `curl`:

```bash
$ wget https://www.google.com/robots.txt
--2020-05-05 08:33:43--  https://www.google.com/robots.txt
Resolving www.google.com (www.google.com)... 172.217.13.164, 2607:f8b0:4020:805::2004
Connecting to www.google.com (www.google.com)|172.217.13.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified [text/plain]
Saving to: ‘robots.txt’

robots.txt                                 [ <=>                                                                        ]   6.84K  --.-KB/s    in 0.001s  

2020-05-05 08:33:43 (9.89 MB/s) - ‘robots.txt’ saved [7004]

$ head -n5 robots.txt
User-agent: *
Disallow: /search
Allow: /search/about
Allow: /search/static
Allow: /search/howsearchworks

> $

Reviewer: Random `$`?

Author: Trailing prompt

The above might be considered a fail since `INDEX,FOLLOW` is the default `<META>` Tag specified by the “Robots Exclusion Protocol” yet `Disallow: /ac.php` is listed in robots.txt.

> ### sitemap.xml

Reviewer: Can we change the naming convention used in the headers? This is simply Sitemaps. I mainly don't enjoy these because of URL issues that might hit by tools that render, and it's actually Sitemaps without the XML extension :yum:

Author: Sure

A sitemap is a file where a developer or organization can provide information about the pages, videos, and other files offered by the site or application, and the relationship between them. Search engines can use this file to more intelligently explore your site. Testers can use `sitemap.xml` files to learn more about the site or application to explore it more completely.

> #### sitemap.xml - with wget

Reviewer: I think this header can simply be dropped ...
